toolkit/tools/pkg/imagecustomizerlib/liveosisobuilder.go

```diff
@@ -324,28 +400,98 @@ func (b *LiveOSIsoBuilder) updateGrubCfg(savedConfigsFilePath string, grubCfgFil
 	if err != nil {
 		return fmt.Errorf("failed to update the kernel arguments with the LiveOS configuration and user configuration in the iso grub.cfg:\n%w", err)
 	}
 
-	err = mergedConfigs.persistSavedConfigs(savedConfigsFilePath)
+	err = file.Write(inputContentString, isoGrubCfgFileName)
 	if err != nil {
-		return fmt.Errorf("failed to save iso configs:\n%w", err)
+		return fmt.Errorf("failed to write %s:\n%w", isoGrubCfgFileName, err)
 	}
 
-	err = file.Write(inputContentString, grubCfgFileName)
+	err = verifyDracutPXESupport(savedConfigs.OS.DracutPackageInfo)
 	if err != nil {
-		return fmt.Errorf("failed to write %s:\n%w", grubCfgFileName, err)
+		logger.Log.Warnf("cannot generate grub.cfg for PXE booting.\n%v", err)
+	} else {
+		err = generatePxeGrubCfg(inputContentString, savedConfigs.Pxe.IsoImageBaseUrl, savedConfigs.Pxe.IsoImageFileUrl,
+			outputImageBase, pxeGrubCfgFileName)
+		if err != nil {
+			return fmt.Errorf("failed to create grub configuration for PXE booting.\n%w", err)
+		}
+	}
+
+	return nil
+}
+
+// generatePxeGrubCfg
+//
+// given the content of the iso grub.cfg, this function derives the PXE
+// equivalent.
+//
+// inputs:
+//   - inputContentString:
+//     iso grub.cfg content.
+//   - pxeIsoImageBaseUrl:
+//     url to a folder containing the iso image to download at boot time.
+//     The function will append the outputImageBase to the url to form the full
+//     url to the image.
+//     For example, if pxeIsoImageBaseUrl is set to "http://192.168.0.1/liveos",
+//     the final url will be "http://192.168.0.1/liveos/<outputImageBase>".
+//     This parameter cannot be set if pxeIsoImageFileUrl is also set.
+//   - pxeIsoImageFileUrl:
+//     url to the iso image to download at boot time.
+//     This parameter cannot be set if pxeIsoImageBaseUrl is also set.
+//   - outputImageBase:
+//     the generated iso name. This value will be used only if the pxeIsoImageFileUrl
+//     is empty.
+//   - pxeGrubCfgFileName:
+//     path of file to hold the PXE grub configuration.
+//
+// returns:
+//   - error: nil if successful, otherwise an error object.
+//
+// generates:
+//   - grub configuration file for PXE booting.
+func generatePxeGrubCfg(inputContentString string, pxeIsoImageBaseUrl string, pxeIsoImageFileUrl string,
+	outputImageBase string, pxeGrubCfgFileName string) error {
+	if pxeIsoImageBaseUrl != "" && pxeIsoImageFileUrl != "" {
+		return fmt.Errorf("cannot set both iso image base url and full image url at the same time.")
+	}
+
+	// remove 'search' commands from PXE grub.cfg because it is not needed.
+	inputContentString, err := removeCommandAll(inputContentString, "search")
+	if err != nil {
+		return fmt.Errorf("failed to remove the 'search' commands from PXE grub.cfg:\n%w", err)
+	}
+
+	// If the specified URL is not a full path to an iso, append the generated
+	// iso file name to it.
+	if pxeIsoImageFileUrl == "" {
+		pxeIsoImageFileUrl, err = url.JoinPath(pxeIsoImageBaseUrl, outputImageBase+".iso")
```

Comment:
I believe this is only correct because `LiveOSIsoBuilder.createIsoImage()` / `isomakerlib` appends `.iso` to the filename. This makes verifying code correctness difficult. I also don't like the fact that the output filename could be different than the one the user provided if the user doesn't use the `.iso` extension.

Reply:
And a minimum, you should at least have a helper function within this file that handles adding `".iso"`. Currently, there are 3 separate locations within this file that does this.

Reply:
Technically, I agree that we should honor the user specified output image name - though it'd be strange if the user wants an iso image that does not have an iso extension. I think, in practice, this will be very rare.